pkg/ingester/flush.go

```diff
@@ -352,87 +357,100 @@ func (i *Ingester) flushChunks(ctx context.Context, fp model.Fingerprint, labelP
 	labelsBuilder.Set(nameLabel, logsValue)
 	metric := labelsBuilder.Labels()
 
-	wireChunks := make([]chunk.Chunk, len(cs))
-
-	// use anonymous function to make lock releasing simpler.
-	err = func() error {
-		chunkMtx.Lock()
-		defer chunkMtx.Unlock()
-
-		for j, c := range cs {
-			// Ensure that new blocks are cut before flushing as data in the head block is not included otherwise.
-			if err := c.chunk.Close(); err != nil {
-				return err
-			}
-			firstTime, lastTime := loki_util.RoundToMilliseconds(c.chunk.Bounds())
-			ch := chunk.NewChunk(
-				userID, fp, metric,
-				chunkenc.NewFacade(c.chunk, i.cfg.BlockSize, i.cfg.TargetChunkSize),
-				firstTime,
-				lastTime,
-			)
-
-			chunkSize := c.chunk.BytesSize() + 4*1024 // size + 4kB should be enough room for cortex header
-			start := time.Now()
-			if err := ch.EncodeTo(bytes.NewBuffer(make([]byte, 0, chunkSize))); err != nil {
-				return err
-			}
-			chunkEncodeTime.Observe(time.Since(start).Seconds())
-			wireChunks[j] = ch
-		}
-		return nil
-	}()
-
-	if err != nil {
-		return err
-	}
-
-	if err := i.store.Put(ctx, wireChunks); err != nil {
-		return err
-	}
-	flushedChunksStats.Inc(int64(len(wireChunks)))
-
-	// Record statistics only when actual put request did not return error.
 	sizePerTenant := chunkSizePerTenant.WithLabelValues(userID)
 	countPerTenant := chunksPerTenant.WithLabelValues(userID)
 
 	chunkMtx.Lock()
 	defer chunkMtx.Unlock()
 
-	for i, wc := range wireChunks {
+	for j, c := range cs {
+		// Ensure that new blocks are cut before flushing as data in the head block is not included otherwise.
+		if err := c.chunk.Close(); err != nil {
+			return fmt.Errorf("chunk close for flushing: %w", err)
+		}
 
-		// flush successful, write while we have lock
-		cs[i].flushed = time.Now()
+		firstTime, lastTime := loki_util.RoundToMilliseconds(c.chunk.Bounds())
+		ch := chunk.NewChunk(
+			userID, fp, metric,
+			chunkenc.NewFacade(c.chunk, i.cfg.BlockSize, i.cfg.TargetChunkSize),
+			firstTime,
+			lastTime,
+		)
 
-		numEntries := cs[i].chunk.Size()
-		byt, err := wc.Encoded()
-		if err != nil {
-			continue
+		if err := i.encodeChunk(ch, c); err != nil {
```

Comment:
Chunk encoding can take time and memory. I'd recommend checking whether `ctx` has timed out before doing it.

Reply:
WDYT of this?
```go
func (i *Ingester) encodeChunk(ctx context.Context, ch chunk.Chunk, desc *chunkDesc) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	start := time.Now()
	chunkBytesSize := desc.chunk.BytesSize() + 4*1024 // size + 4kB should be enough room for cortex header
	if err := ch.EncodeTo(bytes.NewBuffer(make([]byte, 0, chunkBytesSize))); err != nil {
		return fmt.Errorf("chunk encoding: %w", err)
	}
	chunkEncodeTime.Observe(time.Since(start).Seconds())
	return nil
}
```

Reply:
This feels like an unnecessary optimization to me, at most you would save encoding one chunk I believe

Reply:
This allocates up to sightly more than the target chunk size(>1MB by default) and in benchmarks encoding takes 10s of ms per stream so definitely seemed worth the extra 3LOC.